Add Phoenix (Arize) third-party scorer integration

debu-sinha opened the pull request:
### Related Issues/PRs

Partial fix for #19062 (Phoenix portion - TruLens will follow in separate PR)

Split from #19237 per reviewer feedback to separate Phoenix and TruLens integrations.

### What changes are proposed in this pull request?

This PR adds integration for Phoenix (Arize) evaluation framework as MLflow GenAI scorers.

**Phoenix scorers (`mlflow.genai.scorers.phoenix`):**
- `Hallucination`: Detects hallucinations in model outputs
- `Relevance`: Evaluates context relevance to queries
- `Toxicity`: Assesses content toxicity
- `QA`: Evaluates QA correctness
- `Summarization`: Assesses summarization quality

**Implementation details:**
- Lazy loading to avoid import overhead when scorers not used
- Returns `CategoricalRating.YES/NO` values with scores in metadata
- `get_scorer()` API for dynamic metric selection
- Configurable model providers (OpenAI, Databricks, LiteLLM)
- Trace input support for extracting context from retrieval spans
- Clear error messages when optional dependencies are missing

---

## Usage Examples with Real Output

### Direct Scorer Call

```python
from mlflow.genai.scorers.phoenix import Hallucination

scorer = Hallucination(model="openai:/gpt-4o-mini")
feedback = scorer(
    inputs="What is the capital of France?",
    outputs="Paris is the capital of France.",
    expectations={"context": "France is in Europe. Its capital is Paris."},
)
print(f"Value: {feedback.value}")
print(f"Metadata: {feedback.metadata}")
```

**Output:**
```
Value: yes
Metadata: {'mlflow.scorer.framework': 'phoenix', 'score': 0.0, 'label': 'factual'}
```

### With mlflow.genai.evaluate()

```python
import mlflow
from mlflow.genai.scorers.phoenix import Hallucination, Relevance

eval_data = [
    {
        "inputs": {"query": "What is the capital of France?"},
        "outputs": "Paris is the capital of France.",
        "expectations": {"context": "France is in Western Europe. Its capital is Paris."},
    },
    {
        "inputs": {"query": "What is the capital of France?"},
        "outputs": "London is the capital of France.",  # Wrong - should fail
        "expectations": {"context": "France is in Western Europe. Its capital is Paris."},
    },
]

scorers = [
    Hallucination(model="openai:/gpt-4o-mini"),
    Relevance(model="openai:/gpt-4o-mini"),
]

results = mlflow.genai.evaluate(data=eval_data, scorers=scorers)
print(results.metrics)
```

**Output:**
```
{'Relevance/mean': 1.0, 'Hallucination/mean': 0.5}
```

### All 5 Phoenix Scorers

```python
from mlflow.genai.scorers.phoenix import Hallucination, Relevance, Toxicity, QA, Summarization

all_scorers = [
    Hallucination(model="openai:/gpt-4o-mini"),
    Relevance(model="openai:/gpt-4o-mini"),
    Toxicity(model="openai:/gpt-4o-mini"),
    QA(model="openai:/gpt-4o-mini"),
    Summarization(model="openai:/gpt-4o-mini"),
]
results = mlflow.genai.evaluate(data=eval_data, scorers=all_scorers)
```

**Output:**
```
Hallucination/mean: 0.5000
QA/mean: 0.5000
Relevance/mean: 1.0000
Summarization/mean: 0.5000
Toxicity/mean: 1.0000
```

### get_scorer() API

```python
from mlflow.genai.scorers.phoenix import get_scorer

hallucination = get_scorer("Hallucination", model="openai:/gpt-4o-mini")
relevance = get_scorer("Relevance", model="databricks")
toxicity = get_scorer("Toxicity", model="databricks:/my-endpoint")
```

---

### How is this PR tested?

- [x] New unit tests in `tests/genai/scorers/phoenix/test_phoenix.py`
- [x] Real API integration tests with OpenAI (all tests passed)

Test coverage includes:
- All 5 scorers with positive and negative test cases
- Edge cases (empty strings, long text, special characters)
- `get_scorer()` API for all metrics
- `mlflow.genai.evaluate()` batch evaluation
- `CategoricalRating.YES/NO` value validation
- `AssessmentSource.LLM_JUDGE` type validation
- Error handling for missing required fields

### Does this PR require documentation update?

- [x] Yes. I've updated:
  - [x] API references with comprehensive docstrings

### Release Notes

#### Is this a user-facing change?

- [x] Yes

Added Phoenix (Arize) third-party scorer integration for MLflow GenAI evaluation. 5 new scorers available:

**Phoenix scorers:** Hallucination, Relevance, Toxicity, QA, Summarization

Install dependencies:
```bash
pip install arize-phoenix-evals
```

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [x] `area/evaluation`: MLflow model evaluation features

#### How should the PR be classified in the release notes?

- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes

#### Should this PR be included in the next patch release?

- [ ] Yes
- [x] No (this PR will be included in the next minor release)

debu-sinha committed:
Add Phoenix (Arize) third-party scorer integration

Implements support for Phoenix evaluation framework as MLflow GenAI scorers,
enabling seamless integration of Phoenix's LLM evaluation tools within
MLflow's evaluation pipeline.

Phoenix scorers:
- Hallucination: Detects hallucinations in model outputs
- Relevance: Evaluates context relevance to queries
- Toxicity: Assesses content toxicity
- QA: Evaluates QA correctness
- Summarization: Assesses summarization quality

Features:
- Lazy loading to avoid import overhead when not used
- Configurable model providers (OpenAI, Databricks, LiteLLM)
- Trace input support for extracting context from retrieval spans
- Consistent scorer interface returning Feedback objects
- CategoricalRating.YES/NO values with scores in metadata
- get_scorer() API for dynamic metric selection

Install dependencies:
  pip install arize-phoenix-evals

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

B-Step62 commented on mlflow/genai/scorers/phoenix/models.py:
```diff
@@ -0,0 +1,156 @@
+        if provider == "openai":
+            from phoenix.evals import OpenAIModel
+
+            return OpenAIModel(model=model_name)
```
```suggestion
        match provider:
            case "openai":
                from phoenix.evals import OpenAIModel

                return OpenAIModel(model=model_name)
```

B-Step62 commented on mlflow/genai/scorers/phoenix/phoenix.py:
```diff
@@ -0,0 +1,282 @@
+            )
+
+            result = self._evaluator.evaluate(record=record)
+            label, score, explanation = result
```
```suggestion
            label, score, explanation = self._evaluator.evaluate(record=record)
```

B-Step62 reviewed (commented):
Overall looks good! Left a few minor comments

debu-sinha committed:
Address review feedback from @B-Step62

Changes:
- Use call_chat_completions for Databricks managed judge (models.py)
- Use match/case syntax for provider selection (models.py)
- Change @experimental version to 3.9.0 (phoenix.py)
- Replace _positive_label with _label_map for YES/NO/UNKNOWN handling (phoenix.py)
- Remove score clamping - pass through Phoenix scores directly (phoenix.py)
- Keep rationale empty if Phoenix returns no explanation (phoenix.py)
- Add negative_label to registry for label mapping (registry.py)
- Simplify reference extraction with chained or (utils.py)

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha committed:
Refactor Phoenix module structure to match deepeval pattern

- Move PhoenixScorer and metric classes from phoenix.py to __init__.py
- Delete phoenix.py (implementation now in __init__.py)
- Remove docstrings from test functions per MLflow conventions
- Update test mock paths from phoenix.phoenix to phoenix

This matches the module structure used in mlflow/genai/scorers/deepeval/

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

smoorjani commented on mlflow/genai/scorers/phoenix/models.py:
```diff
@@ -0,0 +1,148 @@
+_logger = logging.getLogger(__name__)
+
+
+def _check_phoenix_installed():
```
nit: let's move this into `utils`

smoorjani commented on mlflow/genai/scorers/phoenix/utils.py:
```diff
@@ -0,0 +1,86 @@
+
+    record = {}
+
+    # Parse inputs to string
```
nit: let's remove these one-liners as they don't make the code that much more readable

smoorjani commented on mlflow/genai/scorers/phoenix/utils.py:
```diff
@@ -0,0 +1,86 @@
+
+    # Toxicity evaluator expects text in 'input' field, not 'output'
+    # If only outputs provided and no input, use output as input for Toxicity
+    if metric_name == "Toxicity" and "input" not in record and "output" in record:
```
is it possible to keep metric-specific logic within the metric itself?

smoorjani reviewed (changes requested):
Really excited to see this integration - thanks for doing this work and replicating the existing scorer integration pattern! Left some comments to address

debu-sinha committed:
Address review feedback for Phoenix scorers integration

- Move check_phoenix_installed to utils.py for proper modularization
- Simplify models.py to use LiteLLM for all non-Databricks providers
- Use Phoenix evaluator labels directly as Feedback values
- Add **evaluator_kwargs to allow passing extra params to Phoenix evaluators
- Update expectation key priority: expected_response > context > reference
- Update tests to match new label-based value assertions
- Export get_metric_config from registry for proper test patching

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha committed:
Address review feedback for Phoenix scorer integration

- Simplify utils.py: inline one-liner helpers into main function,
  remove unused metric_name parameter
- Add separate test files following deepeval pattern:
  - test_models.py: tests for Databricks model adapters
  - test_registry.py: tests for evaluator class resolution
  - test_utils.py: tests for input mapping utilities
- Refactor test_phoenix.py:
  - Parameterize scorer tests to reduce duplication
  - Reduce mocking with cleaner fixture approach
  - Remove redundant test cases

All 36 tests pass with full coverage of utils, registry, and models.

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

smoorjani commented on mlflow/genai/scorers/phoenix/models.py:
```diff
@@ -0,0 +1,148 @@
+        self._verbose = False
+
+
+class DatabricksPhoenixModel:
```
followup: let's add a comment explaining why we don't use the base class.

smoorjani commented on tests/genai/scorers/phoenix/test_utils.py:
```diff
@@ -0,0 +1,137 @@
+
+    mock_trace = Mock()
+
+    with (
```
is there a way we can avoid using mocks for this? It'd be great to construct a realistic trace and then test that the record is constructed properly as right now there's no real data being passed into the `map_scorer_inputs_to_phoenix_record` method. same for below.

smoorjani reviewed (changes requested):
thanks for iterating on this - mostly LGTM! just a few smaller follow-ups. Also for the tests, we should make sure they aren't passing right now as we want to rely on the actual phoenix library being called. Assuming they fail due to import errors, we can add in the dependency here: `.github/workflows/master.yml` around L296 similar to https://github.com/mlflow/mlflow/pull/18988/files#diff-38ee08b0d1916cb5b9d8a093e986366eaafd908bc1f516f4fced0033c155d842

debu-sinha commented:
Thanks for the additional feedback @smoorjani! Addressed all comments in the latest commit:

**Changes:**
- Removed duplicate `label` from metadata (already available in `value`)
- Added comment explaining why we don't use Phoenix's `BaseModel` (duck typing for simplicity)
- Simplified registry to only store evaluator class names (removed unused `positive_label`, `negative_label`, `required_fields`)
- Removed `get_metric_config` function (no longer needed)
- Removed one-liner module docstring from utils.py
- Removed standalone export test (imports already tested implicitly in other tests)

All 32 tests pass locally.

debu-sinha commented:
Addressed remaining feedback on test mocking:

**Q: "do we have to mock all of these? generally we should aim to test with minimal mocks"**

Updated the tests to follow the deepeval pattern - now using `pytest.importorskip("phoenix.evals")` instead of `sys.modules` patching. Tests require Phoenix to be installed but only mock the LLM/evaluator calls. This provides better coverage of our integration code while avoiding real API calls.

**Q: "why do we delete these modules?"**

Removed the aggressive module clearing from `test_registry.py`. It was clearing all `mlflow.genai.scorers.phoenix` modules which wasn't necessary - proper imports handle this.

**Additional refactor:**
- Moved `_NoOpRateLimiter` to utils.py per the earlier nit about consolidating Phoenix compatibility helpers.

Tests will be skipped in environments without Phoenix installed, similar to how deepeval tests work.

debu-sinha committed:
Address review feedback on Phoenix scorer integration

Code changes:
- Remove duplicate 'label' from metadata (already in 'value')
- Add comment explaining why not using Phoenix BaseModel
- Simplify registry to only store evaluator class names
- Remove unused get_metric_config function
- Move _NoOpRateLimiter to utils.py

Testing changes:
- Move imports to top-level in test files
- Use real trace objects instead of mocks in test_utils.py
- Add arize-phoenix-evals to CI workflow (master.yml)
- Tests use real Phoenix library with minimal mocking of LLM calls

This approach mirrors how MLflow tests deepeval - the library must be
installed to run tests, but we mock only the actual LLM calls.

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

smoorjani commented on tests/genai/scorers/phoenix/test_models.py:
```diff
@@ -0,0 +1,86 @@
+    create_phoenix_model,
+)
+
+phoenix_evals = pytest.importorskip("phoenix.evals")
```
let's fail directly if we cannot import since not having phoenix installed can hide errors (i.e., just remove this line). this will also skip tests that do not rely on phoenix being installed (e.g., test_databricks_phoenix_model_get_model_name)

smoorjani commented on tests/genai/scorers/phoenix/test_phoenix.py:
```diff
@@ -0,0 +1,192 @@
+            "mlflow.genai.scorers.phoenix.create_phoenix_model",
+            return_value=mock_phoenix_model,
+        ),
+        patch(
```
same here - do we have to mock this? is it possible to mock something more targeted like an LLM call?

smoorjani reviewed (changes requested):
thanks for iterating! almost there, just left a few comments about testing. everything else LGTM

debu-sinha committed:
Address review feedback: remove fixture docstrings and improve test assertions

- Remove one-liner docstrings from test fixtures in test_phoenix.py per
  reviewer request to reduce noise
- Update test_map_scorer_inputs_with_trace to assert exact values instead of
  just checking for key existence, improving test coverage
- Fix type hints for dict parameters to use parameterized types

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha commented:
Addressed all review feedback:

- Added comment in `models.py` explaining why we don't use Arize base class (duck typing for simplicity)
- Removed one-liner docstrings from test fixtures in `test_phoenix.py`
- Updated `test_utils.py` to assert exact values instead of just checking key existence
- Fixed parameterized type hints (`dict[str, str]` instead of `dict`)

All pre-commit hooks pass and tests are green.

@smoorjani @B-Step62 - Ready for re-review when you have a chance. Thanks!

debu-sinha committed:
Merge master into feature/phoenix-scorers and resolve conflicts

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha committed:
Address test review feedback: reduce mocking, use real Phoenix evaluators

- test_models.py: Replace global pytest.importorskip with per-test @pytest.mark.skipif
  so tests that don't need Phoenix (e.g., test_databricks_phoenix_model_get_model_name) run without it
- test_phoenix.py: Use real Phoenix evaluator classes instead of mocking get_evaluator_class
  Only mock create_phoenix_model and evaluator.evaluate (the actual LLM call)
- test_phoenix.py: Add test_phoenix_scorer_evaluator_is_real_instance to verify real evaluators
- test_phoenix.py: Remove evaluator_kwargs test (Phoenix legacy evaluators don't accept kwargs)

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

smoorjani commented on tests/genai/scorers/phoenix/test_phoenix.py:
```diff
@@ -68,124 +50,98 @@ def test_phoenix_scorer(
-            expectations={"expected_response": "test context"},
-        )
+
+        with patch.object(
```
I think this double-nested `with` will fail the CI

smoorjani reviewed (approved):
Conditionally approving on addressing the remaining comments - otherwise, LGTM! Thanks again for iterating on this!

smoorjani commented on .github/workflows/master.yml:
```diff
@@ -286,8 +286,10 @@ jobs:
         run: |
-          source ./dev/install-common-deps.sh
-          pip install openai dspy deepeval litellm ragas arize-phoenix-evals
+          uv sync --extra genai
```
NVM, I was looking at the wrong diff

debu-sinha committed:
Remove conditional test skips - assume packages are installed

Per review feedback: tests should fail if dependencies aren't available,
not skip silently.

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha committed:
Fix CI issues: flatten nested with blocks, remove one-liner docstring

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha committed:
Merge branch 'master' into feature/phoenix-scorers

debu-sinha committed:
Replace pytest.importorskip with direct imports to fail on missing deps

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha commented on mlflow/genai/scorers/phoenix/phoenix.py:
```diff
@@ -0,0 +1,282 @@
+            if score is not None:
+                normalized_score = float(score)
+                # Clamp to 0-1 range if needed
+                if normalized_score < 0.0 or normalized_score > 1.0:
```
Done - passing through original score without clamping.

debu-sinha commented on mlflow/genai/scorers/phoenix/registry.py:
```diff
@@ -0,0 +1,86 @@
+        "required_fields": ["input", "reference"],
+    },
+    "Toxicity": {
+        "evaluator_class": "ToxicityEvaluator",
```
Yes, Phoenix keeps evaluators in the legacy folder. There are newer template-based evaluators but they require more setup. The legacy evaluators are stable and widely used.

debu-sinha commented on tests/genai/scorers/phoenix/test_phoenix.py:
```diff
@@ -0,0 +1,316 @@
+
+    mock_modules = {"phoenix": mock.MagicMock(), "phoenix.evals": mock.MagicMock()}
+    with pytest.raises(MlflowException, match="Unknown Phoenix metric"):
+        with mock.patch.dict("sys.modules", mock_modules):
```
Removed - using real phoenix library directly.

debu-sinha commented on mlflow/genai/scorers/phoenix/registry.py:
```diff
@@ -0,0 +1,85 @@
+_METRIC_REGISTRY = {
+    "Hallucination": {
+        "evaluator_class": "HallucinationEvaluator",
+        "positive_label": "factual",
```
Simplified - now just maps metric name to evaluator class name.

debu-sinha committed:
Fix lint and test failures for Phoenix scorer integration

- Add evaluator_kwargs parameter documentation to get_scorer docstring
- Use walrus operator for span_id_to_context assignment per MLF0048
- Mock litellm.validate_environment in test_create_phoenix_model_openai
- Format test_phoenix.py per ruff formatting requirements

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha committed:
Use monkeypatch.setenv instead of mocking for OpenAI model test

Replace litellm.validate_environment mock with monkeypatch.setenv
to provide the required OPENAI_API_KEY. This approach aligns with
the minimal mocking principle - no mocking, just environment setup.

Signed-off-by: debu-sinha <debusinha2009@gmail.com>

debu-sinha committed:
Merge branch 'master' into feature/phoenix-scorers

B-Step62 reviewed (approved):
LGTM!
